### ISSUE TITLE
Ivariant Violation #31 fired on IE11 - react extension

### ISSUE THREAD

**Issue by dave-flex** [pnp/pnpjs#1161]
Thank you for reporting an issue, suggesting an enhancement, or asking a question. We appreciate your feedback - to help the team understand your
needs please complete the below template to ensure we have the details to help. Thanks!

**Please check out the [Docs](https://pnp.github.io/pnpjs/) to see if your question is already addressed there. This will help us ensure our documentation covers the most frequent questions.**

### Category
- [ ] Enhancement
- [X] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [ **2.0.4** ]

Please specify what version(s) of SharePoint you are targeting: [  **Online**  ]

### Observed Behavior
Extension loads properly on all browsers, but when clicking on edit aat the home page the following error appears on IE11 only:
![error-react](https://user-images.githubusercontent.com/62066601/79372965-c4c04180-7f1b-11ea-8c6b-2801879fb190.png)
going to react link described in the error it says the following:
![image](https://user-images.githubusercontent.com/62066601/79375116-bb37d900-7f1d-11ea-8ffb-45ca52be2bbb.png)

According to some forums the issue was related to polyfill-ie11 version **2.0.1** and needed to downgrade to **1.0.2** but then the page won't load at all:
![noload](https://user-images.githubusercontent.com/62066601/79373424-27194200-7f1c-11ea-92e3-3309fb97e9ec.png)

After some reading and unmet dependencies fixes also attempted to work with pnpjs v**1.3.6** which then reported the following and then went to blank screen:
![error message](https://user-images.githubusercontent.com/62066601/79373509-4a43f180-7f1c-11ea-8c20-539b1a505d3f.png)

Finally removed pnpjs(that's when I identified is an issue with the pnpjs library) and polyfill-ie11 from my code and installed [sp-pnp-js](https://www.npmjs.com/package/sp-pnp-js) which worked with no issue:
![nonpnpjs](https://user-images.githubusercontent.com/62066601/79373734-8e36f680-7f1c-11ea-896d-4d0674102a52.png)

Any help is very much appreciated.

### Steps to Reproduce
create react office ui fabric extension:
@pnp/sp, common, logging, odata, graph v2.0.4 - polyfill-ie11 v2.0.1

apply on communication site and try to edit the page, 
after some seconds of loading the edit section will fail with invariant error #31

Thank you for your feedback!

**Comment by juliemturner** [comment]
Have you added the react.js polyfill that makes react.js work in IE11?

Have you investigated the error using the development as the react.js docs state?

> We highly recommend using the development build locally when debugging your app since it tracks additional debug info and provides helpful warnings about potential problems in your apps, but if you encounter an exception while using the production build, this page will reassemble the original text of the error.

**Comment by patrick-rodgers** [comment]
How does this differ from #1105?

**Comment by dave-flex** [comment]
I have added several imports and search everywhere for an answer, i have added:
import "es6-promise";
import 'whatwg-fetch';
import "react-app-polyfill/ie11";
import "react-app-polyfill/stable";
changed the las 2 mentioned to required, 
attempted with cdn polyfill, and i had no luck to solve this issue, at this point i'm stuck and have nowhere else to go, that's why I thought i might find something direct from here, sorry if I duplicated the issue, it's just confusing how the sp-pnp-js brings no trouble at all to my coding but when i added it the documentation refered me back here since it's deprecated... I'd rather use pnpjs, but at this point I'm overwhelmed and have nowhere else to go.

**Comment by patrick-rodgers** [comment]
sp-pnp-js is the old unmaintained version of what became pnpjs 2.0. It doesn't have any problems with IE11 because we made specific choices/compromises to ensure IE11 was supported. For 2.0 we decided IE11 was no longer a primarily supported browser - though as mentioned in the docs we didn't try and break anyone. It is essentially impossible to support IE11 and use modern JS language features. We did our best to help folks with the polyfill package. My question was simply trying to understand if this was the same issue repeated or something different.

We want to help, but we can't provide on-demand support so it takes time to get to the issues. An option would be to roll your own polyfills to meet your specific needs if ours aren't working, we can't account for all the possible combinations of code.

**Comment by dave-flex** [comment]
Thanks for te reply I'll be pending on any update, or keep you guys posted if I find a solution for the issue, the main problem I've been having is that the polyfills I add are being completly ignored by my code, it asks me for types that does not exists and unfortunately i'm relatively new to the whole environment. 
The only difference I found about 1105 was that i was Using react, as soon as I fixed my react issue I went to the same place as 1105 but the imports that fix 1105 are not working for me.

**Comment by rmbw74** [comment]
@dave-flex  - have you had any luck with this issue, also having problems similar to yours

**Comment by dave-flex** [comment]
@rmbw74  unfortunately no, i had to keep away from pnpjs and use sp-pnp-js(pnp core), which is archived here i believe, just you need to add this 2 polyfills on the very top of your file to avoid issues on IE: 
import "es6-promise";
import "whatwg-fetch";
it could be that you don't have the later installed so you'll need to add it.
I tried with a lot of polyfills, react, ES6, coreJs and so on but it only made it worse.
Since the issue is with IE only my team decided we could be good with that untill my company decides to stop using IE11 and then go back to PNPjs.

Reach out in case I can help :)

**Comment by patrick-rodgers** [comment]
After much thought we determined we don't have a great way to support every possible combination of polyfills folks might need. To that end we have an [updated statement around our polyfills](https://pnp.github.io/pnpjs/concepts/polyfill/#general-statement-on-polyfills). We have included the ability to selectively import just the polyfills you need based on your application. There are just too many different libraries and scenarios for a single package to universally work without conflicts. Thank you to all for the discussion, we appreciate your ideas. If you have continued thoughts please share them in a new issue. Thanks!

**Comment by dave-flex** [comment]
@patrick-rodgers thanks for the info and support 👍
